tests/zfs-tests/tests/functional/cli_root/zfs_mount/zfs_multi_mount.ksh

```diff
@@ -0,0 +1,84 @@
+#!/bin/ksh -p
+#
+# CDDL HEADER START
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy is of the CDDL is also available via the Internet
+# at http://www.illumos.org/license/CDDL.
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright(c) 2018 Datto Inc.
+#
+
+. $STF_SUITE/include/libtest.shlib
+
+#
+# DESCRIPTION:
+# Verify multi mount functionality
+#
+# STRATEGY:
+# 1. Create fs
+# 2. Create and hold open file in filesystem
+# 3. Lazy unmount
+# 4. Verify remounting fs that was lazily unmounted is possible
+# 5. Verify multiple mounts of the same dataset are possible
+# 6. Verify bind mount doesn't prevent rename
+#
+
+verify_runnable "both"
+
+function cleanup
+{
+	log_must pkill -P $TAILPPID tail
+	return 0
+}
+
+log_assert "Verify multiple mounts into one namespace are possible"
+
+log_onexit cleanup
+
+# 1. Create fs
+TESTDS="$TESTPOOL/multi-mount-test"
+log_must zfs create $TESTDS
+MNTPFS="$(get_prop mountpoint $TESTDS)"
+
+# 2. Create and hold open file in filesystem
+FILENAME="$MNTPFS/file"
+log_must mkfile 128k $FILENAME
+log_must tail -f $FILENAME 1>/dev/null 2>/dev/null &
+TAILPPID=$!
+
+# 3. Lazy umount
+MNTPFS2="$MNTPFS-second"
+MNTPFS3="$MNTPFS-third"
+log_must umount -l $MNTPFS
+if [ -f $FILENAME ]; then
+	log_fail "Lazy unmount failed"
+fi
+
+# 4. Verify remounting fs that was lazily unmounted is possible
+log_must zfs mount $TESTDS
+
+# 5. Verify multiple mounts of the same dataset are possible
+log_must mkdir $MNTPFS2
+log_must mount -t zfs -o zfsutil $TESTDS $MNTPFS2
+log_must mkdir $MNTPFS3
+log_must mount -t zfs -o zfsutil $TESTDS $MNTPFS3
+
+# 6. Verify bind mount doesn't prevent rename
+RENAMEFS="$TESTDS-newname"
+MNTPFS4="$MNTPFS-fourth"
+log_must mkdir $MNTPFS4
+log_must mount --bind $MNTPFS $MNTPFS4
+log_must zfs rename $TESTDS $RENAMEFS
+log_must zfs rename $RENAMEFS $TESTDS
+
+log_pass "Multiple mounts are possible"
```

Comment:
In sections 5 and 6 I would want a check for the created files as well (just to be sure).

Reply:
Good idea